pczt/src/sapling.rs

```diff
@@ -0,0 +1,548 @@
+            // Destructure `rhs` to ensure we handle everything.
+            let Spend {
+                cv,
+                nullifier,
+                rk,
+                zkproof,
+                spend_auth_sig,
+                recipient,
+                value,
+                rcm,
+                rseed,
+                rcv,
+                proof_generation_key,
+                witness,
+                alpha,
+                zip32_derivation,
+                dummy_ask,
+                proprietary,
+            } = rhs;
+
+            if lhs.cv != cv || lhs.nullifier != nullifier || lhs.rk != rk {
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.zkproof, zkproof)
+                && merge_optional(&mut lhs.spend_auth_sig, spend_auth_sig)
+                && merge_optional(&mut lhs.recipient, recipient)
+                && merge_optional(&mut lhs.value, value)
+                && merge_optional(&mut lhs.rcm, rcm)
+                && merge_optional(&mut lhs.rseed, rseed)
+                && merge_optional(&mut lhs.rcv, rcv)
+                && merge_optional(&mut lhs.proof_generation_key, proof_generation_key)
+                && merge_optional(&mut lhs.witness, witness)
+                && merge_optional(&mut lhs.alpha, alpha)
+                && merge_optional(&mut lhs.zip32_derivation, zip32_derivation)
+                && merge_optional(&mut lhs.dummy_ask, dummy_ask)
+                && merge_map(&mut lhs.proprietary, proprietary))
+            {
+                return None;
+            }
+        }
+
+        for (lhs, rhs) in self.outputs.iter_mut().zip(outputs.into_iter()) {
+            // Destructure `rhs` to ensure we handle everything.
+            let Output {
+                cv,
+                cmu,
+                ephemeral_key,
+                enc_ciphertext,
+                out_ciphertext,
+                zkproof,
+                recipient,
+                value,
+                rseed,
+                rcv,
+                ock,
+                zip32_derivation,
+                proprietary,
+            } = rhs;
+
+            if lhs.cv != cv
+                || lhs.cmu != cmu
+                || lhs.ephemeral_key != ephemeral_key
+                || lhs.enc_ciphertext != enc_ciphertext
+                || lhs.out_ciphertext != out_ciphertext
+            {
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.zkproof, zkproof)
+                && merge_optional(&mut lhs.recipient, recipient)
+                && merge_optional(&mut lhs.value, value)
+                && merge_optional(&mut lhs.rseed, rseed)
+                && merge_optional(&mut lhs.rcv, rcv)
+                && merge_optional(&mut lhs.ock, ock)
+                && merge_optional(&mut lhs.zip32_derivation, zip32_derivation)
+                && merge_map(&mut lhs.proprietary, proprietary))
```

Comment:
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.